Добавлены UserResponseDto, UserRequestDto и к ним Mapper. Добавлен liquibase

Solyankin opened the pull request.

Commit:
Добавлены UserResponseDto, UserRequestDto и к ним Mapper. Добавлен liquibase

Commit:
Добавил unit тесты на валидацию UserRequestDto

Commit:
Изменил проверку в create_invalid_test

myfrei commented on src/main/java/org/example/controller/user/IUserController.java:
```diff
@@ -0,0 +1,24 @@
+@RestController
+@RequestMapping("/api/v1/users")
+@Tag(name = "User API", description = "Users")
+interface IUserController {
```
давай по стилю Java придерживаться 
interface UserController
реализация UserControllerImpl

Solyankin replied:
Добавляю интерфейс UserController

```java
@RestController
@RequestMapping("/api/v1/users")
@Tag(name = "User API", description = "Users")
interface UserController {
    @GetMapping("/{id}")
    ResponseEntity<UserResponseDto> get(@PathVariable Long id);

    @PostMapping
    ResponseEntity<UserResponseDto> create(@RequestBody UserRequestDto userDto);

    @PutMapping("/{id}")
    ResponseEntity<UserResponseDto> update(@PathVariable Long id, @RequestBody UserRequestDto userDto);

    @DeleteMapping("/{id}")
    ResponseEntity<UserResponseDto> delete(@PathVariable Long id);
}
```
Реализую его в UserControllerImpl, и контроллер исчезает из swagger

```java
@RequiredArgsConstructor
public class UserControllerImpl implements UserController {
    private final UserService service;
    private final UserRequestMapper userRequestMapper;
    private final UserResponseMapper userResponseMapper;

    @Override
    public ResponseEntity<UserResponseDto> get(Long id) {
        Optional<User> user = service.get(id);

        return user
                .map(u -> ResponseEntity.ok(userResponseMapper.toDto(u)))
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @Override
    public ResponseEntity<UserResponseDto> create(UserRequestDto userDto) {
        User user = userRequestMapper.toEntity(userDto);
        User createdUser = service.create(user);
        return ResponseEntity.ok(userResponseMapper.toDto(createdUser));
    }

    @Override
    public ResponseEntity<UserResponseDto> update(Long id, UserRequestDto userDto) {
        User user = userRequestMapper.toEntity(userDto);
        Optional<User> updatedUser = service.update(id, user);

        return updatedUser
                .map(u -> ResponseEntity.ok(userResponseMapper.toDto(u)))
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @Override
    public ResponseEntity<UserResponseDto> delete(Long id) {
        Optional<User> user = service.delete(id);

        return user
                .map(u -> ResponseEntity.ok(userResponseMapper.toDto(u)))
                .orElseGet(() -> ResponseEntity.notFound().build());
    }
}
```

myfrei commented on src/main/java/org/example/controller/user/UserController.java:
```diff
@@ -0,0 +1,63 @@
+
+        return user
+                .map(u -> ResponseEntity.ok(userResponseMapper.toDto(u)))
+                .orElseGet(() -> ResponseEntity.notFound().build());
```
получается логика в контроллере
не по SOLID

myfrei replied:
Для этого отдельный механизм в виде ExeptionHandler

Solyankin replied:
добавил GlobalExceptionHandler

myfrei commented on src/main/java/org/example/controller/contact/model/UserContactResponseDto.java:
```diff
@@ -0,0 +1,14 @@
+import org.example.model.contact.ContactType;
+
+public class UserContactResponseDto {
+
```
не хватает @Data
иными словами getter and setter

myfrei commented on src/main/java/org/example/controller/user/model/UserResponseDto.java:
```diff
@@ -0,0 +1,14 @@
+public class UserResponseDto extends UserRequestDto {
+
+    @Schema(description = "User ID")
+    public Long id;
```
плохая практика отдавать ID из базы наружу
лучше использовать вторичный ключ
а первичный ключ оставлять скрытым от вне

Solyankin replied:
добавил externalId

myfrei commented on src/main/java/org/example/model/user/User.java:
```diff
@@ -20,19 +20,19 @@ public class User {
 
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
```
если используешь ломбок
Можно так же сделать аннотацией

Solyankin replied:
переделал на аннотацию @EqualsAndHashCode(exclude = {"id"})

myfrei commented on src/test/java/org/example/user/UserControllerTest.java:
```diff
@@ -1,102 +1,200 @@
-        User expectedUser = new User(1L, "FirstName", "SecondName");
 
-        when(service.get(expectedUser.getId())).thenReturn(Optional.empty());
+    static Stream<Arguments> validUser() {
```
лучше сделать отдельный класс который будет по SOLID отвечать за модель или метод с пременением before
но статика в тестах плохо

myfrei commented on src/test/java/org/example/user/UserMapperTest.java:
```diff
@@ -0,0 +1,49 @@
+    private UserResponseMapper userResponseMapper;
+
+    @Test
+    void test_response_mapper() {
```
test_response_dto_to_entity_success_map

myfrei commented on src/test/java/org/example/utils/StringUtils.java:
```diff
@@ -0,0 +1,28 @@
+public class StringUtils {
+
+
+    public static String generateRandomString(Integer length) {
```
такие методы уже есть в пакете StringUtils

Commit:
Исправления по ревью:
1. В User добавлены поле externalId и соответствующие изменения в контроллер, сервис и миграции
2. В User добавлены аннотации @RequiredArgsConstructor и @EqualsAndHashCode(exclude = {"id", "externalId"})

3. Класс UserControllerImpl реализует интерфейс UserController

Commit:
Исправления по ревью: добавил GlobalExceptionHandler

Commit:
Fix not found tests

Commit:
Fix not found tests

Commit:
Удалил тесты на валидацию

Commit:
Удалил тесты на валидацию

Commit:
Удалил тесты на валидацию

Commit:
Удалил тесты на валидацию
